### ISSUE TITLE
CA1812 (Avoid uninstantiated internal classes) should not be reported for types passed to a 'new' constraint

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#1158]
#### Analyzer package

Example: Microsoft.CodeQuality.Analyzers

#### Analyzer

CA1812: Avoid uninstantiated internal classes

#### Repro steps

1. Declare an `internal` class
1. Use the class as the argument for a generic type parameter with the `new` constraint

```csharp
internal class InstantiatedType
{
}

internal static class Factory
{
    internal static T Create<T>()
        where T : new()
    {
        return new T();
    }
}

internal class Program
{
    public static void Main(string[] args)
    {
        Console.WriteLine(Factory.Create<InstantiatedType>());
    }
}
```

#### Expected behavior

CA1812 should not be reported.

#### Actual behavior

Unknown; there is no test currently covering this case.

**Comment by 333fred** [comment]
Currently, an error is reported here. Same question as #1163 as to whether we should fix this. /cc @dotnet/roslyn-analysis

**Comment by jinujoseph** [comment]
I think we should fix this.
